File: core-ui/runtime/src/panehost.js
// GoFastr runtime module — PaneHost
//
// Wires the pane-host layout primitive: a primary pane that is always
// visible plus one or two openable side panes (secondary / tertiary).
// Owns the pane lifecycle — open / close / swap, focus handoff on open,
// focus restore on close, and a responsive collapse where, below 768px,
// an open side pane becomes a fixed overlay drawer (backdrop scrim +
// focus trap + scroll lock + ESC-to-close) instead of an inline column.
//
// It does NOT fetch pane content. To fill a pane from a link, use the
// existing data-fui-rpc + data-fui-rpc-signal rail broadcasting into a
// data-fui-signal + data-fui-signal-mode="html" region inside the pane.
//
// Triggers (attribute-driven, delegated):
//   data-fui-pane-open="secondary|tertiary"   open that pane
//   data-fui-pane-close="secondary|tertiary"  close it (bare = topmost)
//   data-fui-pane-swap="secondary|tertiary"   open it, close the sibling
//   data-fui-pane-key="<key>"                 what this trigger opens
// A trigger resolves its host via closest [data-fui-pane-host], or via
// data-fui-pane-host-target="<id>" for triggers outside the host.
//
// URL round-tripping is opt-in per host via data-fui-pane-deeplink
//="<param>". Opening through a keyed trigger writes
// `?<param>=<pane>:<key>`; closing strips it; Back replays the state by
// re-clicking the matching trigger. The server renders first paint from
// the same parameter (ui.PaneDeepLink), so a shared link is not a flash
// of closed pane.
//
// Programmatic API (mirrors openWidget / closeWidget):
//   __gofastr.openPane(hostId, pane)
//   __gofastr.closePane(hostId, pane)
//   __gofastr.swapPane(hostId, pane)
//
// Loads on demand when a [data-fui-pane-host] element appears. The
// breakpoint literal MUST match the @media in the ui-pane-host CSS.
(() => {
  'use strict';
  window.__gofastr = window.__gofastr || {};
  const NS = window.__gofastr;

  const MQ = '(max-width: 768px)';
  const states = new WeakMap();     // host -> per-instance state (Hard Rule 12)
  const overlayHosts = new Set();   // hosts currently in overlay drawer mode

  const SIDE = ['secondary', 'tertiary'];

  // Per-instance state. Multiple pane-hosts per page must not share it.
  // A pane the SERVER rendered open is already visible, so seed the
  // stack from the DOM — otherwise topmost() reports nothing open and
  // ESC, overlay-drawer mode, and bare close all skip a pane the user
  // can plainly see.
  const st = (host) => {
    let s = states.get(host);
    if (!s) {
      s = { stack: [], triggers: {} };
      for (const p of SIDE) {
        const el = paneEl(host, p);
        if (el && !el.hasAttribute('hidden')) s.stack.push(p);
      }
      states.set(host, s);
    }
    return s;
  };
  const paneEl = (host, pane) => host.querySelector('[data-fui-pane="' + pane + '"]');
  const has = (host, pane) => !!paneEl(host, pane);
  const topmost = (host) => { const k = st(host).stack; return k.length ? k[k.length - 1] : null; };
  const sibling = (pane) => (pane === 'secondary' ? 'tertiary' : 'secondary');
  const scrollOwner = (host) => 'panehost:' + (host.id || '');

  function resolveHost(btn) {
    const id = btn.getAttribute('data-fui-pane-host-target');
    if (id) return document.getElementById(id);
    return btn.closest('[data-fui-pane-host]');
  }

  function focusFirst(el) {
    const f = el.querySelector(NS._focusSel);
    const target = f || el;
    if (!f) el.setAttribute('tabindex', '-1'); // make the region itself focusable
    try { target.focus({ preventScroll: true }); } catch (_) { target.focus(); }
  }

  // ─── URL round-tripping (opt-in via data-fui-pane-deeplink) ───────
  //
  // The host names a query parameter; opening through a trigger that
  // declares data-fui-pane-key writes `?<param>=<pane>:<key>`, closing
  // strips it, and Back moves between those states. Pane state is still
  // in-page state (Hard Rule 1) — the parameter only records it so a
  // refresh or a shared link reproduces what is on screen, exactly as
  // widget deep links do. Push/strip both use pushState, matching
  // widgetlinks.js so history behaves the same for panes and modals.
  //
  // syncing suppresses URL writes while we are REACTING to the URL, so
  // replaying history never appends to it.
  let syncing = false;
  const dlParam = (host) => host.getAttribute('data-fui-pane-deeplink');

  function writeURL(mutate) {
    const url = new URL(location.href);
    if (!mutate(url.searchParams)) return;
    const q = url.searchParams.toString();
    history.pushState(null, '', url.pathname + (q ? '?' + q : '') + url.hash);
  }

  function pushPane(host, pane, trigger) {
    const param = dlParam(host);
    if (!param || !trigger) return;
    const key = trigger.getAttribute('data-fui-pane-key');
    // An unkeyed trigger opens the pane without addressing it: there is
    // nothing to put in the URL, so leave it alone.
    if (!key) return;
    const value = pane + ':' + key;
    writeURL((sp) => {
      if (sp.get(param) === value) return false;
      sp.set(param, value);
      return true;
    });
  }

  function stripPane(host, pane) {
    const param = dlParam(host);
    if (!param) return;
    writeURL((sp) => {
      const cur = sp.get(param);
      // Only clear the parameter if it is describing the pane that just
      // closed — never clobber a sibling pane's deep link.
      if (!cur || cur.split(':')[0] !== pane) return false;
      sp.delete(param);
      return true;
    });
  }

  // Bring one host in line with the current URL. Called on popstate.
  function syncFromURL(host) {
    const param = dlParam(host);
    if (!param) return;
    const raw = new URL(location.href).searchParams.get(param) || '';
    const [pane, ...rest] = raw.split(':');
    const key = rest.join(':');
    syncing = true;
    try {
      if (!pane || SIDE.indexOf(pane) < 0 || !has(host, pane)) {
        // No (or unusable) deep link: nothing addressed should be open.
        for (const p of st(host).stack.slice()) closePane(host, p);
        return;
      }
      // Replay the trigger so the pane's CONTENT is restored too — it
      // carries the open marker and the RPC that fills the region, so
      // one click rebuilds the state the URL describes. Falls back to
      // opening an empty pane when the trigger is not on this page.
      const trigger = key &&
        host.querySelector('[data-fui-pane-key="' + CSS.escape(key) + '"]');
      if (trigger) trigger.click();
      else openPane(host, pane, null);
      for (const p of st(host).stack.slice()) {
        if (p !== pane) closePane(host, p);
      }
    } finally {
      syncing = false;
    }
  }

  function openPane(host, pane, trigger) {
    if (!has(host, pane)) return;
    const s = st(host);
    const el = paneEl(host, pane);
    el.removeAttribute('hidden');
    host.classList.add('ui-pane-host--' + pane + '-open');
    if (!s.stack.includes(pane)) s.stack.push(pane);
    if (trigger) s.triggers[pane] = trigger;
    focusFirst(el);
    if (!syncing) pushPane(host, pane, trigger);
    host.dispatchEvent(new CustomEvent('pane-host:open', { bubbles: true, detail: { pane } }));
    syncMode(host);
  }

  function closePane(host, pane) {
    const s = st(host);
    if (!pane) pane = topmost(host);
    if (!pane || !has(host, pane)) return;
    const el = paneEl(host, pane);
    el.setAttribute('hidden', '');
    host.classList.remove('ui-pane-host--' + pane + '-open');
    const i = s.stack.indexOf(pane);
    if (i >= 0) s.stack.splice(i, 1);
    if (!syncing) stripPane(host, pane);
    host.dispatchEvent(new CustomEvent('pane-host:close', { bubbles: true, detail: { pane } }));
    const trig = s.triggers[pane];
    if (trig && typeof trig.focus === 'function') {
      try { trig.focus({ preventScroll: true }); } catch (_) { trig.focus(); }
    }
    delete s.triggers[pane];
    syncMode(host);
  }

  function swapPane(host, pane, trigger) {
    if (!has(host, pane)) return;
    const sib = sibling(pane);
    if (has(host, sib) && st(host).stack.indexOf(sib) >= 0) {
      // A swap is one user action and must leave one history entry, so
      // the sibling's close does not write its own. The open below
      // records the resulting state.
      const was = syncing;
      syncing = true;
      try { closePane(host, sib); } finally { syncing = was; }
    }
    openPane(host, pane, trigger || null);
  }

  // Toggle overlay-drawer mode based on viewport + open state. Sets
  // data-fui-pane-mode on the host (CSS keys the drawer chrome off it)
  // and takes/releases the shared refcounted scroll lock.
  function syncMode(host) {
    const wantOverlay = window.matchMedia(MQ).matches && !!topmost(host);
    const isOverlay = host.getAttribute('data-fui-pane-mode') === 'overlay';
    if (wantOverlay === isOverlay) return;
    if (wantOverlay) {
      host.setAttribute('data-fui-pane-mode', 'overlay');
      overlayHosts.add(host);
      if (NS.doc) NS.doc.lockScroll(scrollOwner(host));
    } else {
      host.removeAttribute('data-fui-pane-mode');
      overlayHosts.delete(host);
      if (NS.doc) NS.doc.unlockScroll(scrollOwner(host));
    }
  }

  // ESC closes; Tab is trapped inside the topmost overlay pane. Minimal
  // trap over NS._focusSel — does NOT touch the widgets module's private
  // _modalStack.
  function onKeydown(e) {
    const host = document.querySelector('[data-fui-pane-mode="overlay"]');
    if (!host) return;
    const pane = topmost(host);
    if (!pane) return;
    if (e.key === 'Escape') { e.preventDefault(); closePane(host, pane); return; }
    if (e.key !== 'Tab') return;
    const el = paneEl(host, pane);
    const items = el.querySelectorAll(NS._focusSel);
    if (!items.length) { e.preventDefault(); return; }
    const first = items[0], last = items[items.length - 1];
    if (e.shiftKey && document.activeElement === first) { e.preventDefault(); last.focus(); }
    else if (!e.shiftKey && document.activeElement === last) { e.preventDefault(); first.focus(); }
  }

  function onClick(e) {
    const trig = e.target.closest(
      '[data-fui-pane-open],[data-fui-pane-close],[data-fui-pane-swap]');
    if (trig) {
      const host = resolveHost(trig);
      if (!host) return;
      if (trig.hasAttribute('data-fui-pane-open')) {
        openPane(host, trig.getAttribute('data-fui-pane-open'), trig);
      } else if (trig.hasAttribute('data-fui-pane-swap')) {
        swapPane(host, trig.getAttribute('data-fui-pane-swap'), trig);
      } else {
        closePane(host, trig.getAttribute('data-fui-pane-close') || topmost(host));
      }
      return;
    }
    // Backdrop click in overlay mode (lands on the host itself) closes.
    const host = e.target.closest('[data-fui-pane-host]');
    if (host && host.getAttribute('data-fui-pane-mode') === 'overlay' && e.target === host) {
      closePane(host, topmost(host));
    }
  }

  // Programmatic API mirroring openWidget / closeWidget.
  function findHost(hostId) {
    return document.getElementById(hostId) ||
      document.querySelector('[data-fui-pane-host]');
  }
  NS.openPane = (hostId, pane) => { const h = findHost(hostId); if (h) openPane(h, pane, null); };
  NS.closePane = (hostId, pane) => { const h = findHost(hostId); if (h) closePane(h, pane); };
  NS.swapPane = (hostId, pane) => { const h = findHost(hostId); if (h) swapPane(h, pane, null); };

  // Single delegated listeners (installed once).
  document.addEventListener('click', onClick);
  document.addEventListener('keydown', onKeydown);

  // One shared matchMedia listener re-syncs every host on viewport change.
  window.matchMedia(MQ).addEventListener('change', () => {
    document.querySelectorAll('[data-fui-pane-host]').forEach(syncMode);
  });

  // Back / forward replays pane state for deep-linked hosts. Hosts
  // without data-fui-pane-deeplink are untouched, so this listener is
  // inert on every existing PaneHost.
  window.addEventListener('popstate', () => {
    document.querySelectorAll('[data-fui-pane-deeplink]').forEach(syncFromURL);
  });

  // Release overlay scroll locks on SPA navigation (Hard Rule 13) so the
  // next page isn't left scroll-locked by a host that's now detached.
  window.addEventListener('gofastr:navigate', () => {
    document.querySelectorAll('[data-fui-pane-host]').forEach((host) => {
      if (host.getAttribute('data-fui-pane-mode') === 'overlay') {
        host.removeAttribute('data-fui-pane-mode');
        if (NS.doc) NS.doc.unlockScroll(scrollOwner(host));
      }
      overlayHosts.delete(host);
    });
  });

  (NS.loadedModules = NS.loadedModules || {}).panehost = true;
  // Idempotent re-wire after SPA-nav: handlers are delegated, so this
  // only ensures state exists for freshly-swapped hosts and re-syncs
  // their initial overlay mode.
  (NS._moduleScanners = NS._moduleScanners || {}).panehost = (root) => {
    const scope = root && root.querySelectorAll ? root : document;
    scope.querySelectorAll('[data-fui-pane-host]').forEach((h) => { st(h); syncMode(h); });
  };
  requestAnimationFrame(() => {
    document.querySelectorAll('[data-fui-pane-host]').forEach((h) => { st(h); syncMode(h); });
  });
})();

File: framework/ui/panehost.go
package ui

// PaneHost is a layout shell with an always-visible primary pane and one
// or two openable side panes (secondary / tertiary). It owns the pane
// LIFECYCLE: show/hide, focus handoff on open, focus restore on close,
// and a responsive collapse where, below 768px, an open side pane
// becomes a fixed overlay drawer (backdrop scrim + focus trap + scroll
// lock + ESC-to-close) instead of an inline grid column.
//
// It does NOT fetch pane content. Loading a link's content into a pane
// uses the EXISTING rails: a trigger carries data-fui-rpc + a
// data-fui-rpc-signal that broadcasts into a data-fui-signal +
// data-fui-signal-mode="html" region inside the pane. Pane open/close
// is in-page state, never a URL route (Hard Rule 1).
//
// A host can still round-trip that state through the URL: set
// DeepLinkParam and the runtime records the open pane in a query
// parameter, so refresh, share, and Back reproduce it. That is a query
// parameter describing in-page state, not a route, the same shape
// widget deep links use for modals.
//
// Shape (mirrors DocLayout, a display:grid whose column count CSS keys
// off the host's open-state hook and the open modifier classes, so no
// inline style is emitted and CSP stays strict):
//
//	list := ui.PaneHost(ui.PaneHostConfig{
//	    Primary:   customerList,
//	    Secondary: detailRegion,   // optional
//	    Tertiary:  inspectorRegion, // optional
//	    SecondaryOpen:  false,      // SSR first-paint state (Hard Rule 6)
//	    SecondaryLabel: "Details",  // labels the role="region"
//	})
//
// See framework/docs/content/pane-host.md for the trigger attributes
// (data-hui-pane-open-control / -close / -swap) and the drawer collapse.

import (
	"net/url"
	"strings"

	"github.com/DonaldMurillo/gofastr/core-ui/html"
	"github.com/DonaldMurillo/gofastr/core-ui/registry"
	"github.com/DonaldMurillo/gofastr/core-ui/style"
	"github.com/DonaldMurillo/gofastr/core/render"
	"github.com/DonaldMurillo/gofastr/framework/headless"
)

// PaneHostConfig configures a PaneHost.
type PaneHostConfig struct {
	// Primary is the always-visible main pane. Required: PaneHost
	// panics when it is empty (mirrors DataTable/DocLayout required
	// slots).
	Primary render.HTML
	// Secondary is the first optional side pane. When empty, no
	// secondary pane is rendered.
	Secondary render.HTML
	// Tertiary is the second optional side pane.
	Tertiary render.HTML

	// SecondaryOpen / TertiaryOpen set the SSR initial open state so
	// the first paint matches server state (Hard Rule 6), e.g. a
	// detail route that should render with the pane already shown. A
	// closed optional pane renders with hidden; the runtime reveals it
	// on open so there is no flash.
	SecondaryOpen bool
	TertiaryOpen  bool

	// SecondaryLabel / TertiaryLabel label each side pane's
	// role="region" via aria-label. Empty falls back to "Secondary" /
	// "Tertiary".
	SecondaryLabel string
	TertiaryLabel  string

	// DeepLinkParam opts this host into URL round-tripping, naming the
	// query parameter that carries pane state, e.g. "pane" for
	// `?pane=secondary:ticket-42`. Opening a pane whose trigger declares
	// a key writes that parameter, closing strips it, and Back moves
	// between those states. Empty (the default) leaves the URL alone.
	//
	// Pane open/close remains in-page state, not a route (Hard Rule 1):
	// the parameter records which pane is showing so a refresh or a
	// shared link reproduces it, exactly as widget deep links do for
	// modals. The server still decides first paint. Read the parameter
	// with PaneDeepLink and set SecondaryOpen/TertiaryOpen plus the
	// pane's content from it. Without that, a shared link renders the
	// URL's pane closed and the runtime opens it after hydration.
	//
	// Triggers declare their key with interactive.PaneKey; a trigger
	// with no key opens the pane without touching the URL.
	DeepLinkParam string

	ID    string
	Class string

	// ExtraAttrs forwards additional attributes (data-* test hooks,
	// analytics markers, ARIA overrides) to the host's root element.
	// Keys the component owns are dropped: class and id (use Class /
	// ID) and data-fui-* (the pane-host marker and deep-link wiring).
	ExtraAttrs html.Attrs
}

// PaneHost renders a primary pane plus one or two openable side panes
// through headless.PaneHost with the fui-pane-host class map (the
// sheet name "ui-pane-host" and marker stay). The root carries the
// open state as data-hui-pane-open (the list the runtime module
// maintains and the sheet's column rules key off, so a client-side
// open changes the columns); the open modifier classes ride along for
// first paint and any caller that reads them.
func PaneHost(cfg PaneHostConfig) render.HTML {
	if cfg.Primary == "" {
		panic("ui: PaneHost requires Primary")
	}

	secondaryOpen := cfg.Secondary != "" && cfg.SecondaryOpen
	tertiaryOpen := cfg.Tertiary != "" && cfg.TertiaryOpen

	rootCls := "fui-pane-host"
	if secondaryOpen {
		rootCls += " fui-pane-host--secondary-open"
	}
	if tertiaryOpen {
		rootCls += " fui-pane-host--tertiary-open"
	}
	if cfg.Class != "" {
		rootCls += " " + cfg.Class
	}

	classes := headless.Classes{
		headless.PartRoot: rootCls,
		headless.PartPane: "fui-pane-host__pane",
	}
	// Each pane's slot modifier rides beside the shared class, resolved
	// by headless.PaneHost through Classes.Variant(PartPane, slot).
	classes[headless.Part("pane--secondary")] = "fui-pane-host__pane--secondary"
	classes[headless.Part("pane--primary")] = "fui-pane-host__pane--primary"
	classes[headless.Part("pane--tertiary")] = "fui-pane-host__pane--tertiary"

	out := headless.PaneHost(headless.PaneHostProps{
		Primary:        cfg.Primary,
		Secondary:      cfg.Secondary,
		Tertiary:       cfg.Tertiary,
		SecondaryOpen:  cfg.SecondaryOpen,
		TertiaryOpen:   cfg.TertiaryOpen,
		SecondaryLabel: cfg.SecondaryLabel,
		TertiaryLabel:  cfg.TertiaryLabel,
		DeepLinkParam:  cfg.DeepLinkParam,
		ID:             cfg.ID,
		ExtraAttrs:     headless.Safe(cfg.ExtraAttrs, "class"),
	}, classes)
	return paneHostStyle.WrapHTML(out)
}

// PaneDeepLink reads a PaneHost deep link out of a request's query.
//
// The value is `<slot>` or `<slot>:<key>`, where slot is "secondary" or
// "tertiary" and key identifies what the pane is showing. It is the
// server half of PaneHostConfig.DeepLinkParam, and it parses exactly
// what the runtime writes. Keep the two in step:
//
//	q := appui.QueryFromContext(ctx)
//	slot, key, ok := ui.PaneDeepLink(q, "pane")
//	detail := emptyState
//	if ok && slot == "secondary" {
//	    if t, found := ticketByID(key); found {
//	        detail = renderTicket(t)   // first paint already shows it
//	    }
//	}
//
// ok is false when the parameter is missing or names something that is
// not an openable side pane, so an edited URL degrades to the ordinary
// closed-pane render rather than an error. Callers must still treat key
// as untrusted input and look it up rather than reflecting it.
//
// Only the first colon splits, so keys may contain colons.
func PaneDeepLink(q url.Values, param string) (slot, key string, ok bool) {
	if param == "" || q == nil {
		return "", "", false
	}
	raw := q.Get(param)
	if raw == "" {
		return "", "", false
	}
	slot, key, _ = strings.Cut(raw, ":")
	// "primary" is always visible and never openable, so it is not a
	// valid deep-link target either.
	if slot != "secondary" && slot != "tertiary" {
		return "", "", false
	}
	return slot, key, true
}

var paneHostStyle = registry.RegisterStyle("ui-pane-host", paneHostCSS)

func paneHostCSS(_ style.Theme) string {
	return `[data-fui-comp="ui-pane-host"] {
  display: grid;
  grid-template-columns: minmax(0, 1fr);
  gap: var(--fui-pane-host-gap, var(--spacing-lg, 16px));
  align-items: start;
  position: relative;
}
[data-fui-comp="ui-pane-host"] .fui-pane-host__pane { min-width: 0; }
/* hidden must win over any pane display rule so a closed pane never
   claims a grid track or paints at first paint. */
[data-fui-comp="ui-pane-host"] [data-hui-pane][hidden] { display: none; }

/* Column count keys off the open-state HOOK the runtime module
   maintains ([data-hui-pane-open], a space-separated list, matched
   token-wise with ~=), so a pane opened in the browser changes the
   columns; the open modifier classes ride along as the first-paint
   spelling. Never inline style (strict CSP, Hard Rule 9b). */
[data-fui-comp="ui-pane-host"].fui-pane-host--secondary-open,
[data-fui-comp="ui-pane-host"][data-hui-pane-open~="secondary"] {
  grid-template-columns: minmax(0, 1fr) var(--fui-pane-host-secondary-w, 360px);
}
[data-fui-comp="ui-pane-host"].fui-pane-host--tertiary-open:not(.fui-pane-host--secondary-open),
[data-fui-comp="ui-pane-host"][data-hui-pane-open~="tertiary"]:not([data-hui-pane-open~="secondary"]) {
  grid-template-columns: minmax(0, 1fr) var(--fui-pane-host-tertiary-w, 300px);
}
[data-fui-comp="ui-pane-host"].fui-pane-host--secondary-open.fui-pane-host--tertiary-open,
[data-fui-comp="ui-pane-host"][data-hui-pane-open~="secondary"][data-hui-pane-open~="tertiary"] {
  grid-template-columns: minmax(0, 1fr) var(--fui-pane-host-secondary-w, 360px) var(--fui-pane-host-tertiary-w, 300px);
}

[data-fui-comp="ui-pane-host"] .fui-pane-host__pane--secondary,
[data-fui-comp="ui-pane-host"] .fui-pane-host__pane--tertiary {
  background: var(--color-surface, transparent);
  border: 1px solid var(--color-border, rgba(0, 0, 0, 0.10));
  border-radius: var(--radii-md, 8px);
  padding: var(--spacing-md, 8px);
}

/* Narrow viewport: the grid collapses to a single column. An open side
   pane renders as a fixed overlay drawer when the module sets
   data-hui-pane-mode="overlay" on the host (it does so once
   matchMedia(max-width: 768px) matches AND a pane is open). The
   breakpoint literal here MUST match the MQ in
   framework/headless/panehost.js. */
@media (max-width: 768px) {
  [data-fui-comp="ui-pane-host"],
  [data-fui-comp="ui-pane-host"].fui-pane-host--secondary-open,
  [data-fui-comp="ui-pane-host"].fui-pane-host--tertiary-open,
  [data-fui-comp="ui-pane-host"].fui-pane-host--secondary-open.fui-pane-host--tertiary-open,
  [data-fui-comp="ui-pane-host"][data-hui-pane-open~="secondary"],
  [data-fui-comp="ui-pane-host"][data-hui-pane-open~="tertiary"],
  [data-fui-comp="ui-pane-host"][data-hui-pane-open~="secondary"][data-hui-pane-open~="tertiary"] {
    grid-template-columns: minmax(0, 1fr);
  }
}

/* Drawer chrome while an open pane is in overlay mode. */
[data-hui-pane-mode="overlay"] [data-hui-pane="secondary"]:not([hidden]),
[data-hui-pane-mode="overlay"] [data-hui-pane="tertiary"]:not([hidden]) {
  position: fixed;
  inset-block: 0;
  inset-inline-end: 0;
  block-size: 100dvh;
  inline-size: min(90vw, var(--fui-pane-host-drawer-w, 420px));
  z-index: var(--z-modal, 300);
  border-radius: 0;
  border-inline-start: 1px solid var(--color-border, rgba(0, 0, 0, 0.10));
  box-shadow: var(--shadow-md, 0 10px 30px rgba(0, 0, 0, 0.18));
  overflow: auto;
}

/* Backdrop scrim while a drawer is open. The host's ::before covers the
   viewport; a click on it lands on the host itself, which the module
   treats as a dismiss (backdrop-click closes the topmost pane). It sits
   one step BELOW the drawer's modal tier: on the framework token scale
   --z-popover (400) is ABOVE --z-modal (300), so the scrim must key off
   --z-modal, not --z-popover, or it would paint over the drawer and eat
   every click inside it. */
[data-hui-pane-mode="overlay"]::before {
  content: "";
  position: fixed;
  inset: 0;
  background: var(--fui-pane-host-scrim, rgba(10, 9, 15, 0.42));
  z-index: calc(var(--z-modal, 300) - 1);
}`
}
